### PR TITLE
Tramp support, fixes #84

### LINKED ISSUE #84: could eglot work with remote project?
hi,
Say a project is on a remote raspberry pi, I have emacs installed on local desktop. When the project file is opened by the desktop emacs (say using tramp), can I get the open file indexed? 

Specificly, the project is a cpp project, the toolchain is only availabe on the remote pi. So the lsp server is better running on the remote pi. Can I get it work by eglot? 

Thanks!

### PR DESCRIPTION
**Pull request by FelipeLema**
fixes #84 

beware: this is only supported in Emacs 27 onwards. I may have used things from nightly, but I'm not certain about it

Pending items:
- [ ] [this pending review](https://github.com/joaotavora/eglot/pull/463#pullrequestreview-408654430) that has already been solved
- [ ] [this only works on Emacs 27, not Emacs 26](https://github.com/joaotavora/eglot/pull/463#issuecomment-664202727). Should we add a [backwards-compatibility layer](https://github.com/joaotavora/eglot/pull/463#issuecomment-665078785) ?

### CODE REVIEW

**Comment by FelipeLema** [comment]
I don't know where to look at in order to fix the checks

**Comment by garyo** [comment]
It would be super cool if this, or a followup to it, would enable using Windows Emacs with a Linux/WSL project (toolchain running in WSL). Currently it doesn't work because of the pathname differences; the WSL toolchain sees "/c/my/project/foo.ts" but Emacs sees "c:/my/project/foo.ts". Because of this I run an extra copy of my toolchain on Windows just for LSP. Even if this PR only supports tramp mode, maybe an extension to it would help me. Interested in your thoughts, @FelipeLema .

**Comment by joaotavora** [comment]
let's do it, I like the patch, but we have to conditionalize this on Emacs-27.  And clean up the commit messages etc.  And get someone to test. I suppose you would @garyo ?

File: eglot.el [hunk 1]
```diff
@@ -1001,6 +1007,9 @@ CONNECT-ARGS are passed as additional arguments to
                  :name (format "autostart-inferior-%s" name)
                  :stderr (format "*%s stderr*" name)
                  :noquery t
+                 :file-handler (file-remote-p
+                                ;; ↓ set at `eglot--connect'
```
Comment (joaotavora) [inline eglot.el:1011]: no unnecessary multibyte chars in source code, please

File: eglot.el [hunk 2]
```diff
@@ -1131,14 +1140,22 @@ If optional MARKER, return a marker instead"
...
   (url-hexify-string
-   (concat "file://" (if (eq system-type 'windows-nt) "/") (file-truename path))
+   (concat "file://" (if (eq system-type 'windows-nt) "/")
+           (tramp-file-local-name (file-truename path)))
    url-path-allowed-chars))
 
-(defun eglot--uri-to-path (uri)
-  "Convert URI to a file path."
+(cl-defun eglot--uri-to-path (uri &optional (server (eglot--current-server-or-lose)))
+  "Convert URI to a file path.
```
Comment (joaotavora) [inline eglot.el:1148]: The first line of the docstring should clear up a doubt I have.

Is it: "`Convert URI to a string pointing to a file in SERVER's host.`"?

If so please use that line only, no more lines needed.

File: eglot.el [hunk 3]
```diff
@@ -718,7 +719,11 @@ be guessed."
                      ((null guess)
                       (format "[eglot] Sorry, couldn't guess for `%s'!\n%s"
                               managed-mode base-prompt))
-                     ((and program (not (executable-find program)))
+                     ((and program
+                           (not
+                            (executable-find program
+                                             (if (file-remote-p default-directory)
```
Comment (joaotavora) [inline eglot.el:725]: You don't need the `if` I think, do you?

**Comment by garyo** [comment]
Yes, I'll try it -- not a tramp expert, and tramp isn't needed for Windows->WSL (just filename mapping) but it would be cool to see it working.

**Comment by FelipeLema** [comment]
will add unit tests next, yo

**Comment by FelipeLema** [comment]
@garyo I don't think tramp support will help you in the short term, because this will only handle the prefix of remote files, but to `clangd` (or `ccls`?) it will still be reported as a "Windows slashed path" since it was built that way. Emacs Windows build handles path differently, and this is where the problem lies.

This being said, this Pull Request will show us _where_ to modify the code to support "Windows slashed paths".

I don't know if these "Windows slashed paths" are part of the LSProtocol and/or you can ask a LSP server if they report URIs this way or if these "Windows slashed paths" are exclusive to `clangd`. Will look into it on the side. Maybe the fix will be included in this PR, although it doesn't look it's going to be this way.

File: eglot.el [hunk 4]
```diff
@@ -77,6 +77,7 @@
 (defvar company-backends)
 (defvar company-tooltip-align-annotations)
 
+(declare-function tramp-file-local-name "tramp" t t)
```
[the remark this answers is not in the document]
Comment (FelipeLema) [reply eglot.el]: you're right... I forgot about it

**Comment by FelipeLema** [comment]
huh? travis CI wasn't triggered?

File: eglot.el [hunk 5]
```diff
@@ -1001,6 +1004,9 @@ CONNECT-ARGS are passed as additional arguments to
                  :name (format "autostart-inferior-%s" name)
                  :stderr (format "*%s stderr*" name)
                  :noquery t
+                 :file-handler (file-remote-p
+                                ;; `default-directory' is set at `eglot--connect'
```
Comment (joaotavora) [inline eglot.el:1008]: You don't need this comment here. `default-directory` affects other stuff (namely `:command`) and users of `make-process` already know that.

**Comment by FelipeLema** [comment]
I activated this PR to get Travis CI to compile, although this is not working in my current setup (hint: it should). Now it just waits forever for an answer from the server.

I wanted to activate Travis to check output from the tests. Maybe get something more helpful than "eglot timed out".

**Comment by joaotavora** [comment]
@FelipeLema don't worry much about Travis.  The important thing is to have tests running with `make check` not necessarily with Travis.

So we need elisp tests that check tramp-related scenarios in `eglot-tests.el`.  

Although, it's certainly nice to have them passing in Travis, if you feel you're getting somewhere. But just so you understand the priorities, Travis CI isn't one of them.  It's likely that Eglot's tests suite will move from Travis CI to something else soon.

File: Makefile [hunk 6]
```diff
@@ -43,6 +43,7 @@ eglot-check: compile
 		-l eglot						\
 		-l eglot-tests						\
 		--eval '(setq ert-batch-backtrace-right-margin 200)'	\
+		--eval '(setq default-directory (concat "/ssh:localhost:" default-directory))'	\
```
Comment (joaotavora) [inline Makefile:46]: Why do we need this? We don't want to run all the tests through a loopback ssh Tramp connection. We just want to do that maybe, for a couple (or more) of tests that are related to Tramp. What if the user running the tests doesn't have the ssh program?

File: eglot-tests.el [hunk 7]
```diff
@@ -69,7 +69,11 @@ then restored."
 
 (defun eglot--call-with-fixture (fixture fn)
   "Helper for `eglot--with-fixture'.  Run FN under FIXTURE."
-  (let* ((fixture-directory (make-temp-file "eglot--fixture" t))
+  (let* ((temporary-file-directory ;; ensure tmp directory respects remote
```
Comment (joaotavora) [inline eglot-tests.el:72]: Similar comment here, though maybe this makes sense.  However, it is only in the subset of Tram tests that we need to worry about these things.  So a second macro like `eglot--with-tramp-fixure` or just directly let-binding this in the new Tramp-related tests, would be sufficient.

File: Makefile (hunk 6 above)
Comment (FelipeLema) [reply Makefile:46]: we don't _need_ this: I'm leveraging Travis to run the tests for me (setups several LSP servers). I wanted to keep this as draft PR, but Travis stopped working.

see my comments above as I've explained

File: eglot-tests.el (hunk 7 above)
Comment (FelipeLema) [reply eglot-tests.el:72]: no need to do a second macro, simply `let`-ing `default-directory` to something remote (ie: with tramp prefix) before calling this function should do the trick.

Comment (joaotavora) [reply eglot-tests.el:72]: > no need to do a second macro, simply let-ing default-directory to something remote (ie: with tramp prefix) before calling this function should do the trick.

Yes, that works fine too.  That's what I meant by "directly let-binding this" in the new tests.

**Comment by FelipeLema** [comment]
it seems that `jsonrpc--process-filter` is not being called at all (in my setup). Running the tests through Travis seems to confirm this.

Is it because the process is not outputting the raw bytes (waiting for newline)? I thought this would've been taken care of by Tramp.

File: Makefile (hunk 6 above)
Comment (joaotavora) [reply Makefile:46]: So what you're saying is that your idea is to run _all_ the tests in a Tramp setting?

**Comment by FelipeLema** [comment]
I don't see any `stty raw` in `tramp-sh.el`

hmmm

**Comment by FelipeLema** [comment]
so far I've tried:
- adding `-tt` to `ssh` command in tramp (through `tramp-methods`)
- `stdbuf -i0 -o0 -e0 clangd` as command (instead of plain `clangd`)

and `clangd` _still_ doesn't answer the initialize method. Don't know how to proceed from here: Tramp is one of those packages with long stateful functions, so it looks pretty bad right now.

**Comment by trittweiler** [comment]
Please make `eglot--apply-workspace-edit` take a server parameter. And then pass down the server argument from the methods for `workspace/applyEdit` and `java.apply.workspaceEdit`, and the local `server` variables in all the other call sites of `eglot--apply-workspace-edit`.

**Comment by trittweiler** [comment]
(I am playing with this right now. I will try to come up with a test case.)

**Comment by trittweiler** [comment]
> it seems that `jsonrpc--process-filter` is not being called at all. Running the tests through Travis seems to confirm this.
> 
> Is it because the process is not outputting the raw bytes (waiting for newline)? I thought this would've been taken care of by Tramp.

One thing that I noticed is that the initial jsonrpc message does not actually make it through to the process on the other end.

I need to add an explicit `\n` in jsonrpc.el's `jsonrpc-connection-send`: `finally return (format "%s\r\n%s\n"` to make it get through. This may need tweakage of the content-length, or it should be `\r\n` rather than just `\n`. I have not read the protocol specification.

**Comment by muffinmad** [comment]
I didn't follow the thread closely but maybe this will help:

After `process-send-string` in the `jsonrpc-connection-send` do the following

```elisp
(when (process-tty-name (jsonrpc--process connection))
  (process-send-eof (jsonrpc--process connection)))
```

Was facing similar issue communicating with `hg` process over TRAMP in vc-hgcmd.

**Comment by FelipeLema** [comment]
@muffinmad it seems your suggestion ends the connection, although I've not debugged it thoroughly

**Comment by FelipeLema** [comment]
@muffinmad also, what's the background on your suggestion? any leads might help here

**Comment by joaotavora** [comment]
> it seems your suggestion ends the connection, although I've not debugged it thoroughly

That doesn't seem too suprising :stuck_out_tongue_winking_eye: it's an `eof` after all. @trittweiler you can try that "against the rails": if it passes Eglot's tests _and_ jsonrpc's self-tests (found in the Emacs repo tree) it shouldn't be too off.

**Comment by muffinmad** [comment]
From [EOT wiki page](https://en.wikipedia.org/wiki/End-of-Transmission_character):

> In Unix, the end-of-file character (by default EOT) causes the terminal driver to make available all characters in its input buffer immediately

Let's see it in action with `cat` process:

```elisp
(setq process-connection-type t)
(setq process-buffer (generate-new-buffer "*process buffer*"))
(setq process (start-file-process "cat process" process-buffer "cat"))
(display-buffer process-buffer)
(process-send-string process "foo")
```

Right now the attentive observer will see no text in the `*process buffer*`. To make the text appear one can send `"\n"`. But in this case the `cat` process will also receive the new line. So here comes `^D` to help:

```elisp
(process-send-string process "\004") ;; The `process-send-eof' do almost the same
```

Of course the great power of `^D`  can't be used here and there: it will terminate the process with no input.

**Comment by trittweiler** [comment]
Thanks @muffinmad . That helped me go down a nice little rabbit hole.

When you make process through the tramp file handler, a process running `ssh ... sh -i` (assuming tramp method is set to `"ssh"`) is created. And in that shell, `exec <command> <args...>` is then run, e.g. `exec ccls`.

Importantly, that process is created with a `connection-type` of `pts` and not `pipe`. Which is why the input is line-buffered and also why `process-send-eof` essentially flushes out the input buffer if there is something in the input buffer. (If there is no input waiting in the input buffer, an actual EOF is sent, terminating the other end. This makes it kind of delicate.)

Another thing is that the stderr output of the process is redirected into a temporary file on the remote system and only retrieved to the local system and then placed into the :stderr buffer when the command finishes.

**Comment by trittweiler** [comment]
I think trying to support this use case ("start language server on the remote system as a pseudo-inferior-process") is more troublesome than it is worth it.

I think, we should support this solely by having the user run the language server on the remote host and export it over a socket, and eglot connects to the socket. We then still need to make sure that the `uri` arguments are translated correctly back and forth. I haven't tested this yet, but will focus on getting this use case right.

(Language servers that do not support communication via socket, can be made so by `socat`. Secure communication can still be established via ssh tunnelling.)

**Comment by trittweiler** [comment]
What I am concerned about is that there's some kind of magic size such that `(process-send-string p <string of size N>)` makes the input be sent to the process, such that a following `(process-send-eof p)` will actually send EOF not EOT because the input buffer was actually empty.

**Comment by joaotavora** [comment]
> I think trying to support this use case ("start language server on the remote system as a pseudo-inferior-process") is more troublesome than it is worth it.

I think we should keep trying: Tramp support is very desirable and doing it with so very few lines of code as @FelipeLema  is showing us even more so.  If `jsonrpc.el` needs to be fixed to detect and work with line-buffering it'd be a worthy investment.  We can also contact Michael Albinus for assistance over at `emacs-devel`.

> I think, we should support this solely by having the user run the language server on the remote host and export it over a socket, and eglot connects to the socket. We then still need to make sure that the uri arguments are translated correctly back and forth. I haven't tested this yet, but will focus on getting this use case right.

Some language servers don't talk over sockets.  And even if they did, that would force us to know the exact syntax to tell each server's executable to start a server on port X, and this varies from server to server.  To be clear, I think know exactly where this suggestion is coming from and it works very well in SLIME/SLY :-)  but the crucial difference is that in that case you're dealing with an evaluation environment for a known language (CL) and you have portability shims.

**Comment by trittweiler** [comment]
> If `jsonrpc.el` needs to be fixed to detect and work with line-buffering it'd be a worthy investment. We can also contact Michael Albinus for assistance over at `emacs-devel`.

You can only send 4095 characters (+ the EOT makes it 4096) to a subprocess with `connection-type` of `pty`. Any additional character is just swallowed.

See the below test function to demonstrate the behaviour.

So, yes, we would need to change jsonrpc to send its payload in suitably-sized chunks.

```emacs-lisp
(defun trittweiler:test-subprocess (connection-type)
  (let* ((buffer (with-current-buffer (get-buffer-create "*cat*")
                   (erase-buffer)
                   (current-buffer)))
         (proc (make-process :command '("cat")
                             :name "cat"
                             :buffer buffer
                             :stderr nil
                             :coding 'utf-8-emacs-unix
                             :connection-type connection-type))
         (+num-chars+ 10000))
    (dotimes (i +num-chars+)
      (process-send-string proc "x"))
    (when (eq connection-type 'pty)
      (process-send-eof proc))
    (cl-assert (accept-process-output proc 1 nil t))
    (let ((num-recv (length (with-current-buffer buffer (buffer-string)))))
      (message "Tried to write %d characters at once, subprocess received %d"
               +num-chars+ num-recv)
      (cons num-recv +num-chars+))))


(trittweiler:test-subprocess 'pipe) ; => (10000 . 10000)
(trittweiler:test-subprocess 'pty)  ; => (4095  . 10000)
```

**Comment by joaotavora** [comment]
> So, yes, we would need to change jsonrpc to send its payload in suitably-sized chunks.

Great! Can you show a patch?  Maybe we should bring this to emacs-devel@gnu.org, can one of you who is most familiar with the problem do that?

**Comment by FelipeLema** [comment]
@trittweiler does this happen with a remote process? both local and remote process? just local?

**Comment by trittweiler** [comment]
> @trittweiler does this happen with a remote process? both local and remote process? just local?

This happens with any process whose `connection-type` is `pty`. The connection that tramp opens (and funnels the remote execution of a program through) is such a `pty` process.

**Comment by joaotavora** [comment]
> >    So, yes, we would need to change jsonrpc to send its payload in suitably-sized chunks.

> Great! Can you show a patch? Maybe we should bring this to emacs-devel@gnu.org, can one of you who is most familiar with the problem do that?

Actually, this is not so.  It appears we do not need to change `jsonrpc` to use chunks.  To  solve the 4095 byte limit problem, it seems to be sufficient to bind `tramp-process-connection-type` to `pipe` around the call to `make-process`.

However, this still doesn't solve the line-buffering problem, which is a different problem. To solve that we probably do need to change it to send a trailing newline always.  For reference, here's the test program I used to discover this:

```
(defvar catty-echo-process nil)
;; (kill-process catty-echo-process)

(defun catty-echo (beg end)
  (interactive
   (if (region-active-p)
       (list (region-beginning) (region-end))
     (error "region must be active")))
  (let* ((buffer (get-buffer-create "*catty-echo*"))
	 (tramp-process-connection-type 'pipe)
         (proc (if (process-live-p catty-echo-process)
                   catty-echo-process
                 (setq catty-echo-process
                       (make-process :command '("cat")
                                     :name "catty-echo"
                                     :buffer buffer
                                     :stderr nil
                                     :coding 'utf-8-emacs-unix
                                     :file-handler (file-remote-p default-directory)
                                     :connection-type 'pipe))))
         (str (buffer-substring-no-properties beg end)))
    (with-current-buffer buffer
      (goto-char (point-max))
      (insert-before-markers
       (format "\n----About to insert %s characters----\n"
               (length str))))
    (process-send-string proc (concat str "\n----CATTYDONE----\n"))
    (catch 'done
      (while t
	(when (not (process-live-p proc)) (error "oops, died"))
	(accept-process-output proc 30)
	(with-current-buffer buffer
	  (save-excursion
	    (goto-char (point-max))
	    (when (looking-back "----CATTYDONE----\n$")
	      (throw 'done nil)))))
      (error "dang"))
    (display-buffer buffer)))
```

**Comment by FelipeLema** [comment]
Per the docs, `tramp-process-connection-type` and consequently `process-connection-type` values are ignored and then bound to `'pipe` when you provide a non-`nil` `:stderr` argument (which we're already doing in eglot).

**Comment by joaotavora** [comment]
> Per the docs, tramp-process-connection-type and consequently process-connection-type values are ignored and then bound to 'pipe when you provide a non-nil :stderr argument (which we're already doing in eglot).

Right, :+1: thanks for the info, that matches my tests too. 

 I got a bit frustrated with this over the weekend. I couldn't confirm, but I think tramp will mix both stdout and stderr in the process filter even if you give if `stderr`.  Also tramp hates it that the `stderr` buffer you give it isn't empty to start with, so jsonrpc.el has to probably stop doing that.

**Comment by FelipeLema** [comment]
will do a fork from here just to sort out the process communication. I'll try to setup a TCP connections on remote machine, `LocalForward` them to local machine (as `ssh` parameter) and leverage the host-port feature of eglot.

It doesn't look like using a remote `make-process` will work out.

**Comment by joaotavora** [comment]
> It doesn't look like using a remote make-process will work out.

Why do you say that? Have you contacted Michael Albinus, the Tramp developer?  Please do so, and include emacs-devel@gnu.org and me in the mail.

**Comment by FelipeLema** [comment]
I wrote to tramp-devel@gnu.org, and Michael Albinus has provided assistance.

Unfortunately, I haven't been able to provide resources (a self-contained test case) for him to do proper testing.

**Comment by joaotavora** [comment]
> I wrote to tramp-devel@gnu.org, and Michael Albinus has provided assistance.

Please, @FelipeLema, CC me and others in that conversation.  I don't subscribe to tramp-devel

**Comment by joaotavora** [comment]
I've found the conversation and CC'ed most of you.  If anyone else is interested in following see https://lists.gnu.org/archive/html/tramp-devel/2020-06/msg00007.html

**Comment by FelipeLema** [comment]
it's currently throwing an error at a very low level call, `insert-file-contents` (under `(tramp-handle-insert-file-contents "/ssh:localhost:/tmp/tramp.z6U8Wg" nil nil nil nil)`). The error is raised fileio.c in emacs source and it isn't very clear what's wrong.

From what I can make out of it, it's trying to insert the content of a directory? if so, why?

PS=I'm following this error here first so we can make something out (considering we have eglot knowledge to track the problem down) before approaching Michael Albinus @ the Emacs / Trame mailing list.

**Comment by joaotavora** [comment]
@FelipeLema , I must admit I'm completely lost in this issue.  You have many commits, but the actual changes to `eglot.el` should be fairly straightfoward.  Most of the work you needed to do is now done, I think.  We just need to pinpoint the technical problems, but I'm having trouble following you.

I think you should roll back the changes to `.travis.yml`. In other words, stop pursuing this strategy which I think is making Travis run all its existing test by TRAMP-connecting to itself before every test.  It is an interesting testing strategy to try later on, maybe.

Instead, my suggestion is to try to describe a simple scenario that everybody can reproduce, giving us:

- instructions on how to load the newest TRAMP sources;
- a specific language server program, and its version;
- a sample directory tree of files (probably just a single sample file will suffice);
- a reliable method for TRAMP-finding one of those files;
- instructions on what exactly do to or type to trigger the same problems or successes that you are experiencing;

**Comment by FelipeLema** [comment]
I've reverted the travis and makefile changes I added for experimenting with tramp.

I'll add a smoke test for Tramp. I think it will be `auto-detect-running-server`, but modified to use a remote directory instead.

**Comment by joaotavora** [comment]
> I'll add a smoke test for Tramp. I think it will be auto-detect-running-server, but modified to use a remote directory instead.

I think this is a good idea.  For now, only one test to debug this behaviour.

**Comment by FelipeLema** [comment]
with https://github.com/joaotavora/eglot/pull/463/commits/10b7a899e045dedc5ac0074a222da394a607e5d8 I can successfully use eglot on a tramp buffer.

Just keeping things clear: the `stty` command was inspired by https://github.com/emacs-lsp/lsp-mode/commit/e17d0c1684caa9349e0a70a8d4045b9df98508e9. I think we don't need copyright assignment since it's one line and it's not an exact copy.

I've also added a test that it's supposed to run when user is standing on a remote directory. I'm missing proper instructions to run such atest, but I don't know if it's worth now.

**Comment by joaotavora** [comment]
@FelipeLema that is good news and progress.

What I would do right now is to bring back a bit of your old Travis technique that you reverted recently.  But only for a single test. Get a single `pyls` test to run over TRAMP by self-connecting to `\ssh:localhost:\<whatever>` but with your new `stty` technique.  Make sure that test passes _and all other tests pass_.  Then we can cleanup the code from there.  If the `file-handler` approach is broken, it's broken. :man_shrugging: 

And yes, I don't think you should worry about the copyright if it's just the one line.  However the commit message should state where you got the idea from.

**Comment by FelipeLema** [comment]
I'm not exactly sure what happened with my setup, but I'm back to the "inserted-chars" error.

This happens when passing a buffer to the `:stderr`  parameter of `make-process. I've tried several workarounds (`stdbuf`) without success, so I'm disabling stderr _only_ on remote processes (by redirecting err to `/dev/null`). I don't think stderr in remote processes is something needed right now.

Still working on the unit test, BTW. It's gonna take me a few days to translate the setup to elisp.

File: eglot.el [hunk 8]
```diff
@@ -875,15 +881,54 @@ This docstring appeases checkdoc, that's all."
...
+                                            tramp-methods))
+                            vec) ;; forward-declare, set by `with-parsed-tramp-file-name'
+                        (when (file-remote-p default-directory)
+                          ;; ensure a pty in ssh command by adding "-tt"
+                          (with-parsed-tramp-file-name (expand-file-name default-directory) vec
+                            (when (string-equal "ssh"
+                                                (tramp-get-method-parameter
+                                                 vec
+                                                 'tramp-login-program))
+                              (setf
```
Comment (FelipeLema) [inline eglot.el:896]: with this `setf` I wanted to do "add `("-tt")` to the list that's inside of `(tramp-get-method-parameter vec 'tramp-login-args)`".

It turned out rather verbose and I believe there must be a succinct way to do it, but I don't know how to.

**Comment by FelipeLema** [comment]
yooooo, I gotta update `eldoc` related code now

**Comment by joaotavora** [comment]
> yooooo, I gotta update eldoc related code now

The current eglot.el master works fine with the current eldoc 1.1.0, to the extent that I can tell (I've only checked briefly).  This is because eldoc 1.1.0 is backward compatible.

I am, however going to change Eglot to the new eldoc now, so you don't have to worry either.

I must say I find your way of working with git very curious.  You're well into the tens of commits here, and they give me an idea of the problems you're facing, but perhaps not as much as a message explaining in plain english.  But of course do continue working as you prefer.

**Comment by FelipeLema** [comment]
Branch-driven development gives you the freedom of doing small commits. I believe small commits are better since they tend to expose _what's going on inside the developer's head_ when they were doing those changes. This is very important later on when you have to maintain code, as you get a better grip of the choices being made when translating an abstract concept to a technical implementation. At the same, they're not meant to be formal documentation. We have separate tools for that (issue tracking, this Pull Request, function doc-strings, READMEs, unit tests) that have their own guidelines to comply to with (overall goal, technical perks, do's and don'ts).

On the matter of documentation, should I add documentation for tramp on the README file? a separate `README.tramp` file? wiki?

**Comment by joaotavora** [comment]
I completely disagree with you (the "inside your head" is helpful for you only). But there's absolutely no problem in that! :-D Keep on working the way you see fit.  Pretty soon I'll be joining in this issue myself.

**Comment by FelipeLema** [comment]
I believe it is now ready to go. @joaotavora , please re-review ( https://github.com/joaotavora/eglot/pull/463#pullrequestreview-410116054 ).

I may have missed something. Let me know if so and I'll fix it

**Comment by nemethf** [comment]
Hi Felipe.  I won't be able to review this PR thoroughly or even try it out.  Additionally, I won't use this feature.  Nevertheless, I'd like to see this merged.  However, it will make my life a bit harder, because I rely on the x-files LSP extension (implemented in eglot-x.el) that allows us to work on tramp files using a local LSP server.  But your approach is definitely better for larger projects.  This PR interferes with eglot-x, but do not worry about that.  I'll fix elgot-x when necessary. 

Regarding the documentation in README.md, I don't see why users should configure anything if clangd is in the users' (remote) path.  Can the tramp support work out-of-the box for simple cases? 

I don't understand this comment: "activate managed mode on other buffers in the same machine this server is running in".  Why do you need special treatment for remote files?  What happens if you have an emacs session with two projects on the same remote machine but with different LSP servers, or an independently opened remote file not belonging to any of those projects?

**Comment by edkolev** [comment]
What's this eglot-x thing, I don't see such a file in the project?

@nemethf I'd advise against supporting so many features in eglot - is there any value in allowing multiple options to work over tramp? Would be confusing for the user, hard for the maintainer

**Comment by edkolev** [comment]
@FelipeLema I tested your branch on emacs 26.3 - I get an error for wrong number of args passed to `executable-find`

I assume emacs 27 has signature `(executable-find COMMAND &optional REMOTE)`
but my emacs has `(executable-find COMMAND)`

trace:
```
Debugger entered--Lisp error: (wrong-number-of-arguments (1 . 1) 2)
  executable-find("gopls" t)
  (if (file-remote-p default-directory) (executable-find program t) (executable-find program))
  (not (if (file-remote-p default-directory) (executable-find program t) (executable-find program)))
  (and program (not (if (file-remote-p default-directory) (executable-find program t) (executable-find program))))
...
```

**Comment by FelipeLema** [comment]
@edkolev yes, you're right. `executable-find` changed signature in version 27: the `&optional REMOTE` was added there. As far as I know, this feature is not available as an (m)ELPA package.

Perhaps this feature can be backported into eglot (like `eglot--executable-find COMMAND &optional REMOTE)`, but I don't know if it's part of the goal of eglot.

**Comment by FelipeLema** [comment]
> 
> 
> Hi Felipe. I won't be able to review this PR thoroughly or even try it out. Additionally, I won't use this feature. Nevertheless, I'd like to see this merged. However, it will make my life a bit harder, because I rely on the x-files LSP extension (implemented in eglot-x.el) that allows us to work on tramp files using a local LSP server. But your approach is definitely better for larger projects. This PR interferes with eglot-x, but do not worry about that. I'll fix elgot-x when necessary.

Let me know if I can help of that. I had no idea of this eglot-x project you mention.
> 
> Regarding the documentation in README.md, I don't see why users should configure anything if clangd is in the users' (remote) path. Can the tramp support work out-of-the box for simple cases?
> 
Yes, it will support simple cases where the command (say `clangd`) is in the `PATH` in the remote server. However, when specifying an explicit _remote_ full path [may not be that obvious](https://github.com/emacs-lsp/lsp-mode/issues/1688).

Perhaps it's too aggressive to put it there in the readme / front page and maybe that info should go somewhere else. Would you have somewhere specific in mind?
> I don't understand this comment: "activate managed mode on other buffers in the same machine this server is running in". Why do you need special treatment for remote files? What happens if you have an emacs session with two projects on the same remote machine but with different LSP servers, or an independently opened remote file not belonging to any of those projects?

Let's say you are working on laptop L and you have remote connections to machines A in Armenia, B in Belarus and C in Chile. For simplicity, you start with Emacs with no opened files.

So you start editing a python file `animal.py` @ A, starting `pyls`. No other buffers should be in managed mode.
Then, you open another python file `beads.py` @ B. It should not be managed by `pyls` @ A since `pyls` @A cannot _reach_ files in B. So you end up with `pyls` @ A managing `animal.py` and `pyls` @ B managing `beads.py`.
Say, you open another file `access.py` @ A. This one should be managed by `pyls` @ A and not by `pyls` @ B, right?

The way it works is that tramp uses a _prefix_ for paths, (`/ssh:A/somewhere/animal.py`, `/ssh:B/elsewhere/beads.py`). This prefix has several properties and we're using it to compare if two paths share the same _connection_ (`/ssh:A/` & `/ssh:B/`).
At the same time, when this prefix is `nil` it means that the path corresponds to a local file (in L). Using `equal` we can compare prefixes and it works out all right (`nil` ≠ `/ssh:A` means that "local file" is not using the same connection as "ssh to machine A").

I can't think of a re-phrasing of this comment that can a) be more clear about it b) uses one or two lines. Suggestions are welcome.

**Comment by nemethf** [comment]
> Let me know if I can help of that. I had no idea of this eglot-x project you mention.

Elgot-x is a project of mine.  I mentioned it because its remote server support is the reason I don't follow your work closely.   You shouldn't worry about eglot-x at this point. (Although it would make my life easier later if eglot--path-to-uri only used file-local-name if the server is a tramp remote server.)

> > Regarding the documentation in README.md, I don't see why users should configure anything if clangd is in the users' (remote) path. Can the tramp support work out-of-the box for simple cases?
> 
> Yes, it will support simple cases where the command (say `clangd`) is in the `PATH` in the remote server. However, when specifying an explicit _remote_ full path [may not be that obvious](https://github.com/emacs-lsp/lsp-mode/issues/1688).
> 
> Perhaps it's too aggressive to put it there in the readme / front page and maybe that info should go somewhere else. Would you have somewhere specific in mind?

I don't know.  Maybe stating this is only necessary for explicitly specifying a remote full path is enough. 

> > I don't understand this comment: "activate managed mode on other buffers in the same machine this server is running in". Why do you need special treatment for remote files? What happens if you have an emacs session with two projects on the same remote machine but with different LSP servers, or an independently opened remote file not belonging to any of those projects?
> 
> Let's say you are working on laptop L and you have remote connections to machines A in Armenia, B in Belarus and C in Chile. For simplicity, you start with Emacs with no opened files.  [...]

I think the approach you described is problematic.  This is the correct approach: two files should be served by the same server if they belong to the same project.  They belong to the same project if their `project-root`s are equal.  It is worth noting that project.el can handle remote files, but the whole project must be on a single (remote or local) system.

**Comment by FelipeLema** [comment]
> 
> 
> > Let me know if I can help of that. I had no idea of this eglot-x project you mention.
> 
> Elgot-x is a project of mine. I mentioned it because its remote server support is the reason I don't follow your work closely. You shouldn't worry about eglot-x at this point. (Although it would make my life easier later if eglot--path-to-uri only used file-local-name if the server is a tramp remote server.)
> 
> > > Regarding the documentation in README.md, I don't see why users should configure anything if clangd is in the users' (remote) path. Can the tramp support work out-of-the box for simple cases?
> > 
> > 
> > Yes, it will support simple cases where the command (say `clangd`) is in the `PATH` in the remote server. However, when specifying an explicit _remote_ full path [may not be that obvious](https://github.com/emacs-lsp/lsp-mode/issues/1688).
> > Perhaps it's too aggressive to put it there in the readme / front page and maybe that info should go somewhere else. Would you have somewhere specific in mind?
> 
> I don't know. Maybe stating this is only necessary for explicitly specifying a remote full path is enough.
will update the README accordingly, then.
> 
> > > I don't understand this comment: "activate managed mode on other buffers in the same machine this server is running in". Why do you need special treatment for remote files? What happens if you have an emacs session with two projects on the same remote machine but with different LSP servers, or an independently opened remote file not belonging to any of those projects?
> > 
> > 
> > Let's say you are working on laptop L and you have remote connections to machines A in Armenia, B in Belarus and C in Chile. For simplicity, you start with Emacs with no opened files.  [...]
> 
> I think the approach you described is problematic. This is the correct approach: two files should be served by the same server if they belong to the same project. They belong to the same project if their `project-root`s are equal. It is worth noting that project.el can handle remote files, but the whole project must be on a single (remote or local) system.

I understand where you're coming from. This is actually an optimization to reduce the number of calls to each function in `project-find-functions`. It works with or without this optimization.

However, calls to functions in `project-find-functions` are not cheap when dealing with remote servers. It can take even seconds to call a single one. Even after spending a lot in tinkering (mostly caching) with every of those functions, it would still take a lot of time to a LangServer to load because it had to cover every buffer. Adding this "reduced buffer list" optimization was worth more than all the tinkering I did myself.

**Comment by nemethf** [comment]
> > > > I don't understand this comment: "activate managed mode on other buffers in the same machine this server is running in". Why do you need special treatment for remote files? What happens if you have an emacs session with two projects on the same remote machine but with different LSP servers, or an independently opened remote file not belonging to any of those projects?

> [...] This is actually an optimization to reduce the number of calls to each function in `project-find-functions`. It works with or without this optimization.

Now that I've actually read the code, I think I understand it.  The code below the comment pre-filters `(buffer-list)` to remove buffers that are on different (remote) systems, therefore the `(eglot--maybe-activate-editing-mode)` is an expensive no-op for them. I'd say the comment didn't help to understand this.

What do you think about something like the following comment instead:

    ;; Remove buffers belonging to a different (remote) system.  They are
    ;; surely not part of this project and calling
    ;; `eglot--maybe-activate-editing-mode' for them can be quite
    ;; expensive.

Additionally, the white-space changes right below this change are unnecessary and confusing.

Also, it's not a big deal, but somehow I'd like to avoid non-ascii chars when possible.  Can you replace «» with something simpler?

Thanks.

**Comment by FelipeLema** [comment]
no problem, applying changes

**Comment by edkolev** [comment]
@FelipeLema what's pending to move this PR forward?

If you need help with testing - I can pull the branch and test

**Comment by FelipeLema** [comment]
I've compiled the pending items in [the first post of this PR](https://github.com/joaotavora/eglot/pull/463#issue-415368701)
